File: ashl.Tokenizer/TokenList.cs
using System.Diagnostics.CodeAnalysis;

namespace ashl.Tokenizer;

public sealed class TokenList<T> where T : TokenBase
{
    
    private readonly LinkedList<T> _list = new();
    
    public Exception CreateException(string message, T token)
    {
        return new Exception();
    }
    public void ThrowExpectedInput()
    {
        throw new Exception("Expected Input");
    }
    public T RemoveFront()
    {
        if (Empty())
        {
            ThrowExpectedInput();
        }

        var a = Front();
        _list.RemoveFirst();
        /*if (NotEmpty())
        {

        }*/
        return a;
    }

    public T RemoveBack()
    {
        if (Empty())
        {
            ThrowExpectedInput();
        }

        var a = Back();
        _list.RemoveLast();
        return a;
    }
    
    // public TokenList<T> ExpectFront(TokenType type)
    // {
    //     if (Empty())
    //     {
    //         ThrowExpectedInput();
    //     }
    //
    //     var a = Front();
    //     
    //     if(a.Type != type) throw CreateException("Unexpected input",a);
    //     return this;
    // }
    
    public TokenList<T> ExpectFront(params TokenType[] type)
    {
        if (Empty())
        {
            ThrowExpectedInput();
        }

        var a = Front();
        
        if(!type.Contains(a.Type)) throw CreateException("Unexpected input",a);
        return this;
    }

    // public TokenList<T> ExpectBack(TokenType type)
    // {
    //     if (Empty())
    //     {
    //         ThrowExpectedInput();
    //     }
    //
    //     var a = Back();
    //     
    //     if(a.Type != type) throw CreateException("Unexpected input",a);
    //     return this;
    // }
    public TokenList<T> ExpectBack(params TokenType[] type)
    {
        if (Empty())
        {
            ThrowExpectedInput();
        }

        var a = Back();
        
        if(!type.Contains(a.Type)) throw CreateException("Unexpected input",a);
        return this;
    }

    public T Front()
    {
        if (Empty())
        {
            ThrowExpectedInput();
        }

        return _list.First();
    }

    public T Back()
    {
        if (Empty())
        {
            ThrowExpectedInput();
        }

        return _list.Last();
    }

    public int Size() => _list.Count;

    public TokenList<T> InsertFront(TokenList<T> other)
    {
        while (other.NotEmpty())
        {
            _list.AddFirst(other.Back());
            other.RemoveBack();
        }
        return this;
    }
    
    public TokenList<T> InsertFront(T token)
    {
        _list.AddFirst(token);
        return this;
    }
    
    public TokenList<T> InsertBack(TokenList<T> other)
    {
        while (other.NotEmpty())
        {
            _list.AddLast(other.Front());
            other.RemoveFront();
        }
        return this;
    }
    
    public TokenList<T> InsertBack(T other)
    {
        _list.AddLast(other);
        return this;
    }
    
    public TokenList<T> Clear()
    {
        _list.Clear();
        return this;
    }

    public bool Empty() => Size() == 0;
    public bool NotEmpty() => Size() > 0;

    public TokenList<T> CreateEmpty()
    {
        return new TokenList<T>();
    }
}

File: typescript/src/TokenList.ts
import Token, { ETokenType } from "./token"

export default class TokenList {
    tokens: Token[];

    constructor(tokens?: Token[]){
        this.tokens = tokens ?? [];
    }

    expectFront(...expected: ETokenType[]): TokenList {
        if(!expected.includes(this.front().type)){
            throw new Error(`Unexpected Token`);
        }
        return this;
    }

    expectBack(...expected: ETokenType[]): TokenList {
        if(!expected.includes(this.back().type)){
            throw new Error(`Unexpected Token`);
        }
        return this;
    }

    insertFront(token: Token): TokenList {
        this.tokens.unshift(token);
        return this;
    }

    insertBack(token: Token): TokenList {
        this.tokens.push(token);
        return this;
    }

    removeFront(): Token {
        return this.tokens.shift();
    }

    removeBack(): Token {
        return this.tokens.pop();
    }

    front(): Token {
        if(this.empty()){
            throw new Error("Expected Input");
        }
        return this.tokens[0];
    }

    back(): Token {
        if(this.empty()){
            throw new Error("Expected Input");
        }
        return this.tokens[this.tokens.length - 1];
    }

    notEmpty() {
        return this.tokens.length !== 0;
    }

    empty() {
        return this.tokens.length === 0;
    }
}